### PR TITLE
feat: add `@Unique` decorator, support specifying multiple unique indexes on the same attribute

### LINKED ISSUE #8148: Many composite unique keys on same field
## The schema:

```js
TABLE person:
id (unique)
username (unique)
...

TABLE events:
id (unique)
person_id (foreign key to person.id, unique: person_start, unique: person_end)
event_start (unique: person_start)
event_end (unique: person_end), description
...
```

## What I want to make happen:
I want to have two composite, unique keys. One key will make sure that there are no events that start at the same time involving the same person. The other key will make sure that there are no events that end at the same time involving the same person. 

On a real-life scenario, every person may participate in as many events as they want even if they overlap, as long as they are present in both the start and end of the event. If two events start and / or end at the same time, the same person cannot be in two places at the same time, so they cannot participate in both events.

## What is actually happening
I can create composite keys (by adding unique: 'key_name' in all fields that should create a unique combo). 
But I cannot set more than one unique key in the same field (this case, person_id). Therefore I cannot create unique keys for both event_start and event_end.

## Details
Dialect: mysql is priority, but I would prefer support for all dialects of sequelize
Sequelize version: 3.X.X

### PR DESCRIPTION
**Pull request by ephys**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Closes #8148

This PR adds the `@Unique` attribute decorator, and extends the "unique" model attribute option to support using the same attribute in multiple composite uniques.

This is now possible:

```ts
const User = current.define('User', {
  firstName: {
    type: DataTypes.STRING,
    // this creates 3 uniques:
    // - one only on firstName
    // - one on firstName & lastName
    // - one on firstName & country
    // (obv you wouldn't want to make names unique but this is just a demo)
    unique: [true, 'firstName-lastName', 'firstName-country'],
  },
  lastName: {
    type: DataTypes.STRING,
    unique: 'firstName-lastName',
  },
  country: {
    type: DataTypes.STRING,
    unique: 'firstName-country',
  },
});
```

Or with decorators:

```ts
class User extends Model {
  @Attribute(DataTypes.STRING)
  @Unique
  @Unique('firstName-lastName')
  @Unique('firstName-country')
  // can also be an array:
  // @Unique([true, 'firstName-lastName', 'firstName-country'])
  declare firstName: string;

  @Attribute(DataTypes.STRING)
  @Unique('firstName-lastName')
  declare lastName: string;

  @Attribute(DataTypes.STRING)
  @Unique('firstName-country')
  declare country: string;
}
```

### CODE REVIEW

**Comment by ephys** [comment]
Don't pay attention to the commit history, I forgot to checkout main before working on this so it includes that commits for `@Attribute` & `@Table`. Final code is fine 😅

File: test/integration/associations/belongs-to-many.test.js [hunk 1]
```diff
@@ -3320,42 +3320,6 @@ describe(Support.getTestDialectTeaser('BelongsToMany'), () => {
       expect(ut1).to.have.length(1);
       expect(ut2).to.have.length(1);
     });
-
-    it('create custom unique identifier', async function () {
```
Comment (ephys) [inline test/integration/associations/belongs-to-many.test.js:3324]: redundant test, already exists in unit tests

File: test/integration/model/create.test.js [hunk 2]
```diff
@@ -1175,20 +1175,6 @@ describe(Support.getTestDialectTeaser('Model'), () => {
       }
     });
 
-    it('raises an error if you mess up the datatype', function () {
```
Comment (ephys) [inline test/integration/model/create.test.js:1178]: redundant test, already exists in unit tests

File: test/unit/model/define.test.js [hunk 3]
```diff
@@ -163,7 +245,7 @@ describe(Support.getTestDialectTeaser('Model'), () => {
             type: DataTypes.MY_UNKNOWN_TYPE,
           },
         });
-      }).to.throw('Unrecognized datatype for attribute "bar.name"');
+      }).to.throw('Attribute "bar.name" does not specify its DataType.');
```
Comment (ephys) [inline test/unit/model/define.test.js:248]: it's more likely that someone will forget to specify the datatype than make a typo in their datatype name, so this error makes more sense I think

File: src/decorators/legacy/legacy-decorator-utils.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+export type PropertyOrGetterDescriptor = (
```
Comment (WikiRik) [inline src/decorators/legacy/legacy-decorator-utils.ts:1]: Maybe just call this file `decorator-utils`? Since it's already in the legacy folder

**Review by WikiRik — APPROVED** [review]
Looks good! Only comment is non-blocking

**Review by fzn0x — DISMISSED** [review]

**Review by WikiRik — APPROVED** [review]

**Pull request merged**